File: packages/openloop/desktop-bridge-host/src/protocol.ts
import { createHmac, randomBytes, timingSafeEqual } from 'node:crypto'

export const BRIDGE_PROTOCOL_VERSION = 1 as const
export const MAX_BRIDGE_FRAME_BYTES = 64 * 1024
const BRIDGE_NONCE_BYTES = 32
const BRIDGE_NONCE_SEQUENCE_BYTES = 8
const BRIDGE_NONCE_ENTROPY_BYTES = BRIDGE_NONCE_BYTES - BRIDGE_NONCE_SEQUENCE_BYTES
const MAX_NONCE_SEQUENCE = 0xffff_ffff_ffff_ffffn

const REQUEST_DOMAIN = Buffer.from('openloop.bridge.request.v1\0')
const RESPONSE_DOMAIN = Buffer.from('openloop.bridge.response.v1\0')
const LOWER_HEX_32_BYTES = /^[0-9a-f]{64}$/u

export interface BridgeRequest {
  readonly version: 1
  readonly requestId: string
  readonly launchId: string
  readonly method: string
  readonly payload: unknown
}

export interface AuthenticatedBridgeRequest {
  readonly request: BridgeRequest
  readonly nonce: string
  readonly mac: string
}

interface BridgeError {
  readonly code: string
  readonly message: string
}

export type BridgeResponse =
  | {
    readonly version: 1
    readonly requestId: string
    readonly ok: true
    readonly result: unknown
    readonly error?: never
  }
  | {
    readonly version: 1
    readonly requestId: string
    readonly ok: false
    readonly result?: never
    readonly error: BridgeError
  }

export interface AuthenticatedBridgeResponse {
  readonly response: BridgeResponse
  readonly nonce: string
  readonly mac: string
}

export interface BridgeVerification {
  readonly launchId: string
  readonly secret: Uint8Array
  readonly nonces: NonceReplayGuard
}

/** Create launch-local nonces with an authenticated monotonic sequence prefix. */
export function createBridgeNonceGenerator(
  entropy: Uint8Array = randomBytes(BRIDGE_NONCE_ENTROPY_BYTES),
): () => Uint8Array {
  if (entropy.length !== BRIDGE_NONCE_ENTROPY_BYTES) {
    throw new TypeError(`bridge nonce entropy must contain exactly ${BRIDGE_NONCE_ENTROPY_BYTES} bytes`)
  }
  const suffix = Uint8Array.from(entropy)
  let sequence = 0n
  return () => {
    if (sequence > MAX_NONCE_SEQUENCE) {
      throw new Error('bridge nonce sequence is exhausted')
    }
    const nonce = Buffer.alloc(BRIDGE_NONCE_BYTES)
    nonce.writeBigUInt64BE(sequence)
    nonce.set(suffix, BRIDGE_NONCE_SEQUENCE_BYTES)
    sequence += 1n
    return nonce
  }
}

/** Bounded launch-local sequence window. Claims are synchronous and atomic in Node. */
export class NonceReplayGuard {
  readonly #maximum: number
  readonly #seen: Uint32Array
  #highest: bigint | undefined

  constructor(maximum = 4096) {
    if (!Number.isSafeInteger(maximum) || maximum <= 0) {
      throw new TypeError('bridge nonce cache size must be a positive safe integer')
    }
    this.#maximum = maximum
    this.#seen = new Uint32Array(Math.ceil(maximum / 32))
  }

  claim(nonceHex: string): void {
    if (!LOWER_HEX_32_BYTES.test(nonceHex)) {
      throw new TypeError('bridge nonce is not canonical lowercase hex')
    }
    const sequence = BigInt(`0x${nonceHex.slice(0, BRIDGE_NONCE_SEQUENCE_BYTES * 2)}`)
    const highest = this.#highest
    if (highest === undefined) {
      this.#highest = sequence
    } else if (sequence > highest) {
      const advance = sequence - highest
      if (advance >= BigInt(this.#maximum)) {
        this.#seen.fill(0)
      } else {
        for (let current = highest + 1n; current <= sequence; current += 1n) {
          this.#clear(current)
        }
      }
      this.#highest = sequence
    } else if (highest - sequence >= BigInt(this.#maximum)) {
      throw new Error('bridge nonce replay rejected')
    }

    if (this.#has(sequence)) throw new Error('bridge nonce replay rejected')
    this.#set(sequence)
  }

  #clear(sequence: bigint): void {
    const index = Number(sequence % BigInt(this.#maximum))
    const word = index >>> 5
    this.#seen[word] = (this.#seen[word] ?? 0) & ~(1 << (index & 31))
  }

  #has(sequence: bigint): boolean {
    const index = Number(sequence % BigInt(this.#maximum))
    return ((this.#seen[index >>> 5] ?? 0) & (1 << (index & 31))) !== 0
  }

  #set(sequence: bigint): void {
    const index = Number(sequence % BigInt(this.#maximum))
    const word = index >>> 5
    this.#seen[word] = (this.#seen[word] ?? 0) | (1 << (index & 31))
  }
}

export function canonicalRequestBytes(
  request: BridgeRequest,
  nonce: Uint8Array,
): Uint8Array {
  validateRequest(request)
  if (nonce.length !== BRIDGE_NONCE_BYTES) {
    throw new TypeError('bridge nonce must contain exactly 32 bytes')
  }
  return Buffer.concat([
    REQUEST_DOMAIN,
    nonce,
    uint32(request.version),
    field(request.requestId),
    field(request.launchId),
    field(request.method),
    field(canonicalJson(request.payload)),
  ])
}

export function authenticateBridgeRequest(
  request: BridgeRequest,
  nonce: Uint8Array,
  secret: Uint8Array,
): AuthenticatedBridgeRequest {
  requireSecret(secret)
  return {
    request,
    nonce: bufferView(nonce).toString('hex'),
    mac: requestMac(request, nonce, secret),
  }
}

export function verifyBridgeRequest(
  value: unknown,
  expected: BridgeVerification,
): BridgeRequest {
  const envelope = parseAuthenticatedRequest(value)
  if (envelope.request.launchId !== expected.launchId) {
    throw new Error('bridge launch is not current')
  }
  const nonce = Buffer.from(envelope.nonce, 'hex')
  try {
    const expectedMac = requestMac(envelope.request, nonce, expected.secret)
    if (!constantTimeHexEqual(envelope.mac, expectedMac)) {
      throw new Error('bridge request authentication failed')
    }
  } finally {
    nonce.fill(0)
  }
  expected.nonces.claim(envelope.nonce)
  return envelope.request
}

export function authenticateBridgeResponse(
  response: BridgeResponse,
  nonce: Uint8Array,
  secret: Uint8Array,
): AuthenticatedBridgeResponse {
  validateResponse(response)
  requireSecret(secret)
  if (nonce.length !== BRIDGE_NONCE_BYTES) {
    throw new TypeError('bridge nonce must contain exactly 32 bytes')
  }
  return {
    response,
    nonce: bufferView(nonce).toString('hex'),
    mac: responseMac(response, nonce, secret),
  }
}

export function verifyBridgeResponse(
  value: unknown,
  expected: {
    readonly requestId: string
    readonly nonce: Uint8Array
    readonly secret: Uint8Array
  },
): BridgeResponse {
  const envelope = parseAuthenticatedResponse(value)
  if (envelope.response.requestId !== expected.requestId) {
    throw new Error('bridge response request id does not match')
  }
  const expectedNonce = bufferView(expected.nonce).toString('hex')
  if (!constantTimeHexEqual(envelope.nonce, expectedNonce)) {
    throw new Error('bridge response nonce does not match')
  }
  const expectedMac = responseMac(envelope.response, expected.nonce, expected.secret)
  if (!constantTimeHexEqual(envelope.mac, expectedMac)) {
    throw new Error('bridge response authentication failed')
  }
  return envelope.response
}

export function encodeBridgeFrame(value: unknown): Uint8Array {
  const body = Buffer.from(canonicalJson(value))
  const header = Buffer.allocUnsafe(4)
  try {
    if (body.length === 0 || body.length > MAX_BRIDGE_FRAME_BYTES) {
      throw new Error('bridge frame is oversized or empty')
    }
    header.writeUInt32BE(body.length)
    return Buffer.concat([header, body])
  } finally {
    header.fill(0)
    body.fill(0)
  }
}

export function decodeBridgeFrame(frame: Uint8Array): unknown {
  if (frame.length < 4) throw new Error('bridge frame header is truncated')
  const bytes = Buffer.from(frame.buffer, frame.byteOffset, frame.byteLength)
  const bodyLength = bytes.readUInt32BE(0)
  if (bodyLength === 0 || bodyLength > MAX_BRIDGE_FRAME_BYTES) {
    throw new Error('bridge frame is oversized or empty')
  }
  if (bytes.length !== bodyLength + 4) {
    throw new Error('bridge frame length does not match')
  }
  return JSON.parse(bytes.subarray(4).toString('utf8')) as unknown
}

function canonicalResponseBytes(response: BridgeResponse, nonce: Uint8Array): Uint8Array {
  const body = response.ok
    ? { ok: true, result: response.result }
    : { error: response.error, ok: false }
  return Buffer.concat([
    RESPONSE_DOMAIN,
    nonce,
    uint32(response.version),
    field(response.requestId),
    field(canonicalJson(body)),
  ])
}

function requestMac(
  request: BridgeRequest,
  nonce: Uint8Array,
  secret: Uint8Array,
): string {
  const canonical = canonicalRequestBytes(request, nonce)
  try {
    return createHmac('sha256', secret).update(canonical).digest('hex')
  } finally {
    canonical.fill(0)
  }
}

function responseMac(
  response: BridgeResponse,
  nonce: Uint8Array,
  secret: Uint8Array,
): string {
  const canonical = canonicalResponseBytes(response, nonce)
  try {
    return createHmac('sha256', secret).update(canonical).digest('hex')
  } finally {
    canonical.fill(0)
  }
}

function parseAuthenticatedRequest(value: unknown): AuthenticatedBridgeRequest {
  const envelope = exactRecord(value, ['request', 'nonce', 'mac'], 'bridge request envelope')
  if (typeof envelope.nonce !== 'string' || !LOWER_HEX_32_BYTES.test(envelope.nonce)
    || typeof envelope.mac !== 'string' || !LOWER_HEX_32_BYTES.test(envelope.mac)) {
    throw new TypeError('bridge request authentication fields are invalid')
  }
  const parsed = exactRecord(
    envelope.request,
    ['version', 'requestId', 'launchId', 'method', 'payload'],
    'bridge request',
  )
  const request = parsed as unknown as BridgeRequest
  validateRequest(request)
  return { request, nonce: envelope.nonce, mac: envelope.mac }
}

function parseAuthenticatedResponse(value: unknown): AuthenticatedBridgeResponse {
  const envelope = exactRecord(value, ['response', 'nonce', 'mac'], 'bridge response envelope')
  if (typeof envelope.nonce !== 'string' || !LOWER_HEX_32_BYTES.test(envelope.nonce)
    || typeof envelope.mac !== 'string' || !LOWER_HEX_32_BYTES.test(envelope.mac)) {
    throw new TypeError('bridge response authentication fields are invalid')
  }
  const response = envelope.response as BridgeResponse
  validateResponse(response)
  return { response, nonce: envelope.nonce, mac: envelope.mac }
}

function validateRequest(request: BridgeRequest): void {
  if ((request as { readonly version: unknown }).version !== BRIDGE_PROTOCOL_VERSION) {
    throw new TypeError('bridge protocol version is unsupported')
  }
  if (typeof request.requestId !== 'string' || request.requestId.length === 0
    || typeof request.launchId !== 'string' || request.launchId.length === 0
    || typeof request.method !== 'string' || request.method.length === 0) {
    throw new TypeError('bridge request fields are invalid')
  }
  canonicalJson(request.payload)
}

function validateResponse(response: BridgeResponse): void {
  const fields = response.ok
    ? ['version', 'requestId', 'ok', 'result']
    : ['version', 'requestId', 'ok', 'error']
  exactRecord(response, fields, 'bridge response')
  if ((response as { readonly version: unknown }).version !== BRIDGE_PROTOCOL_VERSION
    || typeof response.requestId !== 'string' || response.requestId.length === 0) {
    throw new TypeError('bridge response fields are invalid')
  }
  if (response.ok) {
    canonicalJson(response.result)
  } else if (typeof response.error.code !== 'string'
    || response.error.code.length === 0
    || typeof response.error.message !== 'string') {
    throw new TypeError('bridge response error is invalid')
  } else {
    exactRecord(response.error, ['code', 'message'], 'bridge response error')
  }
}

function exactRecord(value: unknown, fields: readonly string[], label: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new TypeError(`${label} must be an object`)
  }
  const actual = Reflect.ownKeys(value)
  if (actual.length !== fields.length
    || actual.some(key => typeof key !== 'string' || !fields.includes(key))) {
    throw new TypeError(`${label} fields are invalid`)
  }
  return value as Record<string, unknown>
}

function canonicalJson(value: unknown, ancestors = new Set<object>()): string {
  if (value === null) return 'null'
  if (typeof value === 'boolean') return value ? 'true' : 'false'
  if (typeof value === 'string') {
    if (/[\uD800-\uDFFF]/u.test(value)) {
      throw new TypeError('bridge JSON strings must not contain UTF-16 surrogate code units')
    }
    return JSON.stringify(value)
  }
  if (typeof value === 'number') {
    if (!Number.isSafeInteger(value)) {
      throw new TypeError('bridge JSON numbers must be safe integers')
    }
    return String(value)
  }
  if (typeof value !== 'object') {
    throw new TypeError('bridge values must be canonical JSON')
  }
  if (ancestors.has(value)) throw new TypeError('bridge values must not be cyclic')
  ancestors.add(value)
  try {
    if (Array.isArray(value)) {
      return `[${value.map(entry => canonicalJson(entry, ancestors)).join(',')}]`
    }
    const prototype = Object.getPrototypeOf(value) as object | null
    if (prototype !== Object.prototype && prototype !== null) {
      throw new TypeError('bridge JSON objects must be plain records')
    }
    const record = value as Record<string, unknown>
    const keys = Reflect.ownKeys(record)
    if (keys.some(key => typeof key !== 'string')) {
      throw new TypeError('bridge JSON objects must have string keys')
    }
    const sorted = (keys as string[]).sort((left, right) => (
      Buffer.compare(Buffer.from(left, 'utf8'), Buffer.from(right, 'utf8'))
    ))
    return `{${sorted.map(key => `${canonicalJson(key)}:${canonicalJson(record[key], ancestors)}`).join(',')}}`
  } finally {
    ancestors.delete(value)
  }
}

function field(value: string): Buffer {
  const bytes = Buffer.from(value)
  return Buffer.concat([uint32(bytes.length), bytes])
}

function uint32(value: number): Buffer {
  if (!Number.isSafeInteger(value) || value < 0 || value > 0xffff_ffff) {
    throw new TypeError('bridge canonical field length is out of bounds')
  }
  const bytes = Buffer.allocUnsafe(4)
  bytes.writeUInt32BE(value)
  return bytes
}

function requireSecret(secret: Uint8Array): void {
  if (secret.length < 32) throw new TypeError('bridge secret must contain at least 32 bytes')
}

function constantTimeHexEqual(actual: string, expected: string): boolean {
  if (!LOWER_HEX_32_BYTES.test(actual) || !LOWER_HEX_32_BYTES.test(expected)) return false
  const actualBytes = Buffer.from(actual, 'hex')
  const expectedBytes = Buffer.from(expected, 'hex')
  try {
    return timingSafeEqual(actualBytes, expectedBytes)
  } finally {
    actualBytes.fill(0)
    expectedBytes.fill(0)
  }
}

function bufferView(bytes: Uint8Array): Buffer {
  return Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength)
}

File: apps/openloop-desktop/src-tauri/src/bridge/protocol.rs
use std::{
    io::{self, Read, Write},
    sync::Mutex,
};

use hmac::{Hmac, Mac};
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use serde_json::Value;
use sha2::Sha256;
use uuid::Uuid;
use zeroize::{Zeroize, Zeroizing};

pub const BRIDGE_PROTOCOL_VERSION: u8 = 1;
pub const MAX_BRIDGE_FRAME_BYTES: usize = 64 * 1024;
pub const BRIDGE_NONCE_BYTES: usize = 32;
const MAX_NONCES: usize = 4096;
const MAX_SAFE_INTEGER: u64 = 9_007_199_254_740_991;
const REQUEST_DOMAIN: &[u8] = b"openloop.bridge.request.v1\0";
const RESPONSE_DOMAIN: &[u8] = b"openloop.bridge.response.v1\0";

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BridgeRequest {
    pub version: u8,
    pub request_id: String,
    pub launch_id: String,
    pub method: String,
    pub payload: Value,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct AuthenticatedBridgeRequest {
    pub request: BridgeRequest,
    pub nonce: String,
    pub mac: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct BridgeError {
    pub code: String,
    pub message: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BridgeResponse {
    pub version: u8,
    pub request_id: String,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<BridgeError>,
}

impl BridgeResponse {
    pub fn success(request_id: impl Into<String>, result: Value) -> Self {
        Self {
            version: BRIDGE_PROTOCOL_VERSION,
            request_id: request_id.into(),
            ok: true,
            result: Some(result),
            error: None,
        }
    }

    pub fn failure(
        request_id: impl Into<String>,
        code: impl Into<String>,
        message: impl Into<String>,
    ) -> Self {
        Self {
            version: BRIDGE_PROTOCOL_VERSION,
            request_id: request_id.into(),
            ok: false,
            result: None,
            error: Some(BridgeError {
                code: code.into(),
                message: message.into(),
            }),
        }
    }
}

impl Zeroize for BridgeResponse {
    fn zeroize(&mut self) {
        self.version = 0;
        self.request_id.zeroize();
        self.ok = false;
        if let Some(result) = &mut self.result {
            zeroize_json_value(result);
        }
        self.result = None;
        if let Some(error) = &mut self.error {
            error.code.zeroize();
            error.message.zeroize();
        }
        self.error = None;
    }
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct AuthenticatedBridgeResponse {
    pub response: BridgeResponse,
    pub nonce: String,
    pub mac: String,
}

impl Zeroize for AuthenticatedBridgeResponse {
    fn zeroize(&mut self) {
        self.response.zeroize();
        self.nonce.zeroize();
        self.mac.zeroize();
    }
}

pub struct NonceReplayGuard {
    maximum: usize,
    window: Mutex<NonceReplayWindow>,
}

struct NonceReplayWindow {
    highest: Option<u64>,
    seen: Vec<bool>,
}

impl Default for NonceReplayGuard {
    fn default() -> Self {
        Self::new(MAX_NONCES)
    }
}

impl NonceReplayGuard {
    pub fn new(maximum: usize) -> Self {
        assert!(maximum > 0, "nonce cache must not be empty");
        Self {
            maximum,
            window: Mutex::new(NonceReplayWindow {
                highest: None,
                seen: vec![false; maximum],
            }),
        }
    }

    pub fn claim(&self, nonce: [u8; BRIDGE_NONCE_BYTES]) -> io::Result<()> {
        let sequence = u64::from_be_bytes(
            nonce[..8]
                .try_into()
                .map_err(|_| invalid("bridge nonce sequence is invalid"))?,
        );
        let mut window = self
            .window
            .lock()
            .map_err(|_| invalid("bridge nonce cache lock is poisoned"))?;
        if let Some(highest) = window.highest {
            if sequence > highest {
                let advance = sequence - highest;
                if advance >= self.maximum as u64 {
                    window.seen.fill(false);
                } else {
                    for current in highest + 1..=sequence {
                        window.seen[(current % self.maximum as u64) as usize] = false;
                    }
                }
                window.highest = Some(sequence);
            } else if highest - sequence >= self.maximum as u64 {
                return Err(permission_denied("bridge nonce replay rejected"));
            }
        } else {
            window.highest = Some(sequence);
        }
        let index = (sequence % self.maximum as u64) as usize;
        if window.seen[index] {
            return Err(permission_denied("bridge nonce replay rejected"));
        }
        window.seen[index] = true;
        Ok(())
    }
}

pub fn canonical_request_bytes(
    request: &BridgeRequest,
    nonce: &[u8; BRIDGE_NONCE_BYTES],
) -> io::Result<Vec<u8>> {
    validate_request(request)?;
    let payload = canonical_json(&request.payload)?;
    let mut bytes = Vec::with_capacity(
        REQUEST_DOMAIN.len()
            + nonce.len()
            + 4
            + request.request_id.len()
            + request.launch_id.len()
            + request.method.len()
            + payload.len()
            + 16,
    );
    bytes.extend_from_slice(REQUEST_DOMAIN);
    bytes.extend_from_slice(nonce);
    bytes.extend_from_slice(&(u32::from(request.version)).to_be_bytes());
    put_field(&mut bytes, request.request_id.as_bytes())?;
    put_field(&mut bytes, request.launch_id.as_bytes())?;
    put_field(&mut bytes, request.method.as_bytes())?;
    put_field(&mut bytes, payload.as_bytes())?;
    Ok(bytes)
}

pub fn sign_request(
    request: BridgeRequest,
    nonce: [u8; BRIDGE_NONCE_BYTES],
    secret: &[u8],
) -> io::Result<AuthenticatedBridgeRequest> {
    require_secret(secret)?;
    let mac = hmac_sha256(secret, &canonical_request_bytes(&request, &nonce)?)?;
    Ok(AuthenticatedBridgeRequest {
        request,
        nonce: encode_hex(&nonce),
        mac: encode_hex(&mac),
    })
}

pub fn verify_request<'a>(
    envelope: &'a AuthenticatedBridgeRequest,
    expected_launch_id: &Uuid,
    secret: &[u8],
    nonces: &NonceReplayGuard,
) -> io::Result<&'a BridgeRequest> {
    if envelope.request.version != BRIDGE_PROTOCOL_VERSION {
        return Err(permission_denied("bridge protocol version is unsupported"));
    }
    if envelope.request.launch_id != expected_launch_id.to_string() {
        return Err(permission_denied("bridge launch is not current"));
    }
    require_secret(secret)?;
    let nonce = decode_fixed_hex::<BRIDGE_NONCE_BYTES>(&envelope.nonce, "bridge nonce")?;
    let actual_mac = decode_fixed_hex::<32>(&envelope.mac, "bridge MAC")?;
    verify_hmac_sha256(
        secret,
        &canonical_request_bytes(&envelope.request, &nonce)?,
        &actual_mac,
    )?;
    nonces.claim(nonce)?;
    Ok(&envelope.request)
}

pub fn sign_response(
    response: BridgeResponse,
    nonce: [u8; BRIDGE_NONCE_BYTES],
    secret: &[u8],
) -> io::Result<AuthenticatedBridgeResponse> {
    validate_response(&response)?;
    require_secret(secret)?;
    let canonical = canonical_response_bytes(&response, &nonce)?;
    let mac = hmac_sha256(secret, &canonical)?;
    Ok(AuthenticatedBridgeResponse {
        response,
        nonce: encode_hex(&nonce),
        mac: encode_hex(&mac),
    })
}

pub fn encode_frame<T: Serialize>(value: &T) -> io::Result<Vec<u8>> {
    Ok(encode_frame_zeroizing(value)?.to_vec())
}

fn encode_frame_zeroizing<T: Serialize>(value: &T) -> io::Result<Zeroizing<Vec<u8>>> {
    let body = Zeroizing::new(
        serde_json::to_vec(value)
            .map_err(|error| invalid(format!("bridge frame JSON is invalid: {error}")))?,
    );
    if body.is_empty() || body.len() > MAX_BRIDGE_FRAME_BYTES {
        return Err(invalid("bridge frame is oversized or empty"));
    }
    let length = u32::try_from(body.len()).map_err(|_| invalid("bridge frame is oversized"))?;
    let mut frame = Zeroizing::new(Vec::with_capacity(4 + body.len()));
    frame.extend_from_slice(&length.to_be_bytes());
    frame.extend_from_slice(&body);
    Ok(frame)
}

pub fn write_frame<W: Write, T: Serialize>(writer: &mut W, value: &T) -> io::Result<()> {
    writer.write_all(&encode_frame_zeroizing(value)?)
}

pub fn read_frame<R: Read>(reader: &mut R) -> io::Result<Vec<u8>> {
    let mut header = [0_u8; 4];
    reader.read_exact(&mut header)?;
    let length = u32::from_be_bytes(header) as usize;
    if length == 0 || length > MAX_BRIDGE_FRAME_BYTES {
        return Err(invalid("bridge frame is oversized or empty"));
    }
    let mut body = vec![0_u8; length];
    reader.read_exact(&mut body)?;
    Ok(body)
}

pub fn read_json_frame<R: Read, T: DeserializeOwned>(reader: &mut R) -> io::Result<T> {
    serde_json::from_slice(&read_frame(reader)?)
        .map_err(|error| invalid(format!("bridge frame JSON is invalid: {error}")))
}

fn canonical_response_bytes(
    response: &BridgeResponse,
    nonce: &[u8; BRIDGE_NONCE_BYTES],
) -> io::Result<Zeroizing<Vec<u8>>> {
    validate_response(response)?;
    let mut body = if response.ok {
        serde_json::json!({ "ok": true, "result": response.result.as_ref().unwrap() })
    } else {
        serde_json::json!({ "error": response.error.as_ref().unwrap(), "ok": false })
    };
    let serialized = canonical_json(&body);
    zeroize_json_value(&mut body);
    let body = Zeroizing::new(serialized?);
    let mut bytes = Zeroizing::new(Vec::with_capacity(
        RESPONSE_DOMAIN.len() + nonce.len() + response.request_id.len() + body.len() + 12,
    ));
    bytes.extend_from_slice(RESPONSE_DOMAIN);
    bytes.extend_from_slice(nonce);
    bytes.extend_from_slice(&(u32::from(response.version)).to_be_bytes());
    put_field(&mut bytes, response.request_id.as_bytes())?;
    put_field(&mut bytes, body.as_bytes())?;
    Ok(bytes)
}

fn validate_request(request: &BridgeRequest) -> io::Result<()> {
    if request.version != BRIDGE_PROTOCOL_VERSION
        || request.request_id.is_empty()
        || request.launch_id.is_empty()
        || request.method.is_empty()
    {
        return Err(invalid("bridge request fields are invalid"));
    }
    canonical_json(&request.payload).map(|_| ())
}

fn validate_response(response: &BridgeResponse) -> io::Result<()> {
    if response.version != BRIDGE_PROTOCOL_VERSION || response.request_id.is_empty() {
        return Err(invalid("bridge response fields are invalid"));
    }
    match (&response.ok, &response.result, &response.error) {
        (true, Some(result), None) => validate_json(result),
        (false, None, Some(error)) if !error.code.is_empty() => Ok(()),
        _ => Err(invalid("bridge response result shape is invalid")),
    }
}

fn canonical_json(value: &Value) -> io::Result<String> {
    validate_json(value)?;
    serde_json::to_string(value)
        .map_err(|error| invalid(format!("bridge canonical JSON failed: {error}")))
}

fn validate_json(value: &Value) -> io::Result<()> {
    match value {
        Value::Null | Value::Bool(_) | Value::String(_) => Ok(()),
        Value::Number(number) => {
            let valid = number
                .as_u64()
                .is_some_and(|value| value <= MAX_SAFE_INTEGER)
                || number
                    .as_i64()
                    .is_some_and(|value| value.unsigned_abs() <= MAX_SAFE_INTEGER);
            if valid {
                Ok(())
            } else {
                Err(invalid("bridge JSON numbers must be safe integers"))
            }
        }
        Value::Array(values) => values.iter().try_for_each(validate_json),
        Value::Object(values) => values.values().try_for_each(validate_json),
    }
}

pub(crate) fn zeroize_json_value(value: &mut Value) {
    match value {
        Value::Null => {}
        Value::Bool(boolean) => *boolean = false,
        Value::Number(number) => *number = serde_json::Number::from(0),
        Value::String(string) => string.zeroize(),
        Value::Array(values) => {
            values.iter_mut().for_each(zeroize_json_value);
            values.clear();
        }
        Value::Object(values) => {
            for (mut key, mut nested) in std::mem::take(values) {
                key.zeroize();
                zeroize_json_value(&mut nested);
            }
        }
    }
}

fn put_field(bytes: &mut Vec<u8>, field: &[u8]) -> io::Result<()> {
    let length = u32::try_from(field.len()).map_err(|_| invalid("bridge field is oversized"))?;
    bytes.extend_from_slice(&length.to_be_bytes());
    bytes.extend_from_slice(field);
    Ok(())
}

fn require_secret(secret: &[u8]) -> io::Result<()> {
    if secret.len() < 32 {
        Err(invalid("bridge secret must contain at least 32 bytes"))
    } else {
        Ok(())
    }
}

fn hmac_sha256(secret: &[u8], message: &[u8]) -> io::Result<[u8; 32]> {
    let mut mac = Hmac::<Sha256>::new_from_slice(secret)
        .map_err(|_| invalid("bridge secret cannot initialize HMAC"))?;
    mac.update(message);
    Ok(mac.finalize().into_bytes().into())
}

fn verify_hmac_sha256(secret: &[u8], message: &[u8], actual: &[u8; 32]) -> io::Result<()> {
    let mut mac = Hmac::<Sha256>::new_from_slice(secret)
        .map_err(|_| invalid("bridge secret cannot initialize HMAC"))?;
    mac.update(message);
    mac.verify_slice(actual)
        .map_err(|_| permission_denied("bridge request authentication failed"))
}

fn encode_hex(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

pub(crate) fn decode_nonce(value: &str) -> io::Result<[u8; BRIDGE_NONCE_BYTES]> {
    decode_fixed_hex(value, "bridge nonce")
}

fn decode_fixed_hex<const N: usize>(value: &str, label: &str) -> io::Result<[u8; N]> {
    if value.len() != N * 2
        || !value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
    {
        return Err(invalid(format!("{label} is not canonical lowercase hex")));
    }
    let mut bytes = [0_u8; N];
    for (index, byte) in bytes.iter_mut().enumerate() {
        *byte = u8::from_str_radix(&value[index * 2..index * 2 + 2], 16)
            .map_err(|_| invalid(format!("{label} is invalid")))?;
    }
    Ok(bytes)
}

fn invalid(message: impl Into<String>) -> io::Error {
    io::Error::new(io::ErrorKind::InvalidData, message.into())
}

fn permission_denied(message: impl Into<String>) -> io::Error {
    io::Error::new(io::ErrorKind::PermissionDenied, message.into())
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::zeroize_json_value;

    #[test]
    fn recursive_json_zeroization_scrubs_nested_secret_material() {
        let mut value = json!({
            "secret": [115, 101, 99, 114, 101, 116],
            "nested": {
                "text": "sensitive",
                "flag": true,
            },
            "sensitive-key": "value",
        });

        zeroize_json_value(&mut value);

        assert_eq!(value, json!({}));
    }
}
